src/main/scala/edu/berkeley/cs/rise/opaque/execution/operators.scala

```diff
@@ -308,6 +307,73 @@ case class EncryptedSortMergeJoinExec(
   }
 }
 
+case class EncryptedBroadcastNestedLoopJoinExec(
+    left: SparkPlan,
+    right: SparkPlan,
+    buildSide: BuildSide,
+    joinType: JoinType,
+    condition: Option[Expression])
+    extends BinaryExecNode with OpaqueOperatorExec {
+
+  override def output: Seq[Attribute] = {
+    joinType match {
+      case _: InnerLike =>
+        left.output ++ right.output
+      case LeftOuter =>
+        left.output ++ right.output.map(_.withNullability(true))
+      case RightOuter =>
+        left.output.map(_.withNullability(true)) ++ right.output
+      case FullOuter =>
+        left.output.map(_.withNullability(true)) ++ right.output.map(_.withNullability(true))
+      case j: ExistenceJoin =>
+        left.output :+ j.exists
+      case LeftExistence(_) =>
+        left.output
+      case x =>
+        throw new IllegalArgumentException(
+          s"BroadcastNestedLoopJoin should not take $x as the JoinType")
+    }
+  }
+
+  override def executeBlocked(): RDD[Block] = {
+    val joinExprSer = Utils.serializeJoinExpression(
+        joinType, None, None, left.output, right.output, condition)
+
+    val leftRDD = left.asInstanceOf[OpaqueOperatorExec].executeBlocked()
+    val rightRDD = right.asInstanceOf[OpaqueOperatorExec].executeBlocked()
+
```

Comment:
This is what Spark does but I found the existing implementation much easier and shorter. The issue with (streamed, broadcast) is that I would have to convert LeftAnti to RightAnti, etc. for the C++ code otherwise the result with contain the wrong columns.

Reply:
Correct: left side always corresponds to `outer_rows` and right side always corresponds to `inner_rows` in C++. I think the strongest case for keeping the code as is comes from strategies.scala: Under the current scheme, if we find a significant performance boost for, as an example, building left when the join is `LeftSemi` (could be due to left side being much smaller), then there will be no required changes to operators.scala or the C++ code. If we decide to model after the Spark code, then switching `outer_rows` and `inner_rows` will mean implementing `RightSemi` in C++ even though it is not officially supported by Scala.